functionary/core/utils/workflow.py

```diff
@@ -0,0 +1,103 @@
+from typing import TYPE_CHECKING
+
+from django.db import transaction
+
+from core.models import Workflow, WorkflowStep
+
+if TYPE_CHECKING:
+    from core.models import Function
+
+
+def add_step(
+    workflow: Workflow,
+    name: str,
+    function: "Function",
+    parameter_template: str,
+    next: WorkflowStep | None = None,
+) -> WorkflowStep:
+    """Add a WorkflowStep to the specified point in the Workflow
+
+    Args:
+        workflow: Workflow to add a step to
+        name: Name of the WorkflowStep
+        function: Function to task
+        parameter_template: Template string that will be rendered to form the
+            parameter json for the function
+        next: The WorkflowStep to insert the new step before. The default value
+            of None will insert at the end of the Workflow.
+
+    Returns:
+        The created WorkflowStep
+
+    Raises:
+        ValueError: workflow and next.workflow do not match
+    """
+    if next is not None and workflow != next.workflow:
+        raise ValueError("Provided next step is not part of provided workflow")
+
+    previous_step = WorkflowStep.objects.filter(workflow=workflow, next=next).first()
+
+    with transaction.atomic():
+        new_step = WorkflowStep.objects.create(
+            workflow=workflow,
+            name=name,
+            function=function,
+            parameter_template=parameter_template,
+            next=previous_step.next if previous_step else None,
+        )
+
+        if previous_step:
+            previous_step.next = new_step
+            previous_step.save()
+
+    return new_step
+
+
+def remove_step(step: WorkflowStep) -> None:
+    """Remove a WorkflowStep from a Workflow
+
+    Args:
+        step: The WorkflowStep to remove
+
+    Returns:
+        None
+    """
+    before = WorkflowStep.objects.filter(next=step).first()
+
+    with transaction.atomic():
+        if before:
+            before.next = step.next
+            before.save()
+
+        step.delete()
+
+
+def move_step(step: WorkflowStep, next: WorkflowStep | None = None) -> None:
```

Comment:
Can this function get a more descriptive name like `move_step_before` since the parameters are both the same type?

Reply:
I feel like the parameter names and the docstring make it mostly clear enough so I'm inclined to keep the name as is unless you really feel flip the table over strong on this one.